doc/prebuilt/man/man1/slackspot.1
'\" t
.\"     Title: slackspot
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 2014/06/26
.\"    Manual: Xenomai Manual
.\"    Source: Xenomai 3.0.5
.\"  Language: English
.\"
.TH "SLACKSPOT" "1" "2014/06/26" "Xenomai 3\&.0\&.5" "Xenomai Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
slackspot \- Trace secondary mode switches
.SH "SYNOPSIS"
.sp
\fBslackspot\fR [ options ]
.SH "DESCRIPTION"
.sp
\fBslackspot\fR is a utility to decode the trace data collected by the Cobalt core when CONFIG_XENO_OPT_DEBUG_TRACE_RELAX is enabled in the kernel configuration\&.
.sp
This data describes each call hierarchy causing migration to secondary mode (i\&.e\&. \fIrelaxes\fR) within the application\&. \fBslackspot\fR presents such data in a human\-readable format as symbolic stack backtraces, for helping in debugging spurious relaxes\&.
.SH "OPTIONS"
.sp
\fBslackspot\fR accepts the following options:
.PP
\fB\-\-file <trace\-file>\fR
.RS 4
Read the trace information to decode from
\fItrace\-file\fR\&. By default, trace data is read from
/proc/xenomai/debug/relax
unless the standard input stream was redirected, in which case
stdin
is read\&. In addition, the dash character "\-" is interpreted as a placeholder for
stdin\&.
.RE
.PP
\fB\-\-path <dir[:dir\&...]>\fR
.RS 4
Search directory list for executables and dynamic libraries\&. Directories are separatared by a semi\-colon within the list\&. Each directory may be scanned for binary executables when resolving symbols found in stack backtraces\&.
.RE
.PP
\fB\-\-filter\-in <name=exp[,name=\&...]>\fR
.RS 4
Only retain backtraces matching the given filters in the output\&. Each filter is specified by a
\fIname=<expr>\fR
pair, where
\fIname\fR
identifies the information field to be matched in the backtrace, and
\fIexpr\fR
is a regular expression which should match such data\&. Filters are separated by a comma within the list\&. The available filters are as follows:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIthread\fR
matches the thread name\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIpid\fR
matches the kernel task identifier, i\&.e\&. per\-task
\fIpid\fR\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIexe\fR
matches the name of the main executable being traced\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIfunction\fR
matches the name of the function being traced\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIfile\fR
matches the path of the source file being traced\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fImap\fR
matches the path of the mapped executable being traced\&.
.RE
.RE
.PP
\fB\-\-filter <name=exp[,name=\&...]>\fR
.RS 4
A shorthand for
\fB\-\-filter\-in\fR\&.
.RE
.PP
\fB\-\-filter\-out <name=exp[,name=\&...]>\fR
.RS 4
Only retain backtraces NOT matching the given filters in the output\&. This option inverts the sense of matching defined by
\fB\-\-filter\-in\fR\&.
.RE
.PP
\fBCROSS_COMPILE=<toolchain\-prefix>\fR
.RS 4
A cross\-compilation toolchain prefix should be specified for decoding the data obtained from a target system, on a build/development machine\&. When present, the value of CROSS_COMPILE will be prepended to
gcc
and
addr2line
for running the corresponding utilities on the development system\&.
.RE
.SH "VERSIONS"
.sp
\fBslackspot\fR appeared in Xenomai 3\&.0 for the \fICobalt\fR real\-time core\&.
.SH "EXAMPLE"
.sp
In the following scenario, the \fItarget\fR system built with the CONFIG_XENO_OPT_DEBUG_TRACE_RELAX feature enabled in the kernel configuration, just ran the \fI/bin/relax\fR program\&.
.sp
This program caused a transition to secondary mode switch of the current task (\fITask 2\fR) as a result of calling putchar()\&. The Cobalt core saved the corresponding backtrace information, which is now available from /proc/xenomai/debug/relax on the target system\&.
.sp
Since the target system has limited horsepower, and doesn\(cqt have access to the binary utilities required for decoding the trace data, we will send such data over the network to the \fIhost\fR system, in order for the latter to do the decoding and display the call stacks\&.
.sp
We use the standard netcat utility to send and receive the contents of /proc/xenomai/debug/relax over the wire between the target and host systems\&. The host will also have to mention where the cross\-compilation toolchain can be found, by setting the CROSS_COMPILE variable appropriately\&. The example assumes that /opt/rootfs/MPC5200/lib is the host\-based location of the system libraries mounted over NFS onto the target file hierarchy\&.
.PP
\fBOn the target system:\fR. 
.sp
.if n \{\
.RS 4
.\}
.nf
target> netcat \-l \-p 67676 \-c < /proc/xenomai/debug/relax
.fi
.if n \{\
.RE
.\}
.PP
\fBOn the host system:\fR. 
.sp
.if n \{\
.RS 4
.\}
.nf
host> netcat target 67676 | CROSS_COMPILE=ppc_6xx\- slackspot
      \-\-path=/opt/rootfs/MPC5200/lib:$HOME/frags/relax \-\-filter thread=Task*
Thread[828] "Task 2" started by /bin/relax:
   #0  0xfff00000 ???
   #1  0x000001bb ???
   #2  0x00064393 _IO_file_doallocate() in ??:?
   #3  0x00073d6f _IO_doallocbuf() in ??:?
   #4  0x00072d87 _IO_file_overflow() in ??:?
   #5  0x00075f83 __overflow() in ??:?
   #6  0x0006997b putchar() in ??:?
   #7  0x100017db task2_func() in /home/rpm/frags/relax/relax\&.c:23
   #8  0x000078d7 task_entry() in /home/rpm/git/xenomai\-forge/lib/alchemy/task\&.c:235
   #9  0x00005a6b start_thread() in pthread_create\&.c:?
   #10 0x000d389f __clone() in ??:?
.fi
.if n \{\
.RE
.\}
.sp
.SH "AUTHOR"
.sp
\fBslackspot\fR was written by Philippe Gerum <rpm@xenomai\&.org>\&.


ChangeLog
2009-03-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vxworks/semLib.c (sem_destroy_internal) : Add missing
	rescheduling. Fix sem_read_proc().

2009-03-19  Philippe Gerum  <rpm@xenomai.org>

	* include/asm-generic/wrappers.h, src/nucleus/shadow.c: Fix
	generic bits for 2.6.29.

2009-03-18  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/pipe.c: Fix streaming mode.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/nucleus/shadow.c: Trace syscall exits.

	* patching file include/asm-generic/syscall.h,
	ksrc/nucleus/shadow.c: Add xn_sys_current_info.

	* src/rtdk/rt_print.c: suspend/resume printer thread.

	* ksrc/skins/native/syscall.c, src/skins/native/task.c: Support
	T_WARNSW on task creation.

	* src/skins/native/task.c: Avoid lazy non-TLS rt_task_self.

	* src/skins/native/task.c: Silent migration before self-deletion.

	* src/skins, ksrc/nucleus/shadow.c: Rework initial Linux
	schedparam setting.

	* ksrc/nucleus, include/asm-generic, ksrc/arch/generic: Exclude
	CPUs from Xenomai use.

	* ksrc/nucleus/shadow.c: Cleanup taskexit hook.

	* ksrc/nucleus/shadow.c: Suppress switch warning upon signal
	receipt when the process is being debugged.

	* ksrc/nucleus/sched.c: Allow to change the watchdog timeout via a
	module parameter.

	* ksrc/arch/x86/usercopy_{32, 64}.c: Fix miscompilation of
	__do_strncpy_from_user.

	* src/rtdk, include/rtdk.h: Add assert_context helpers.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/skins/posix/syscall.c: Do not return -EFAULT on zero-length
	names. Instead, return -EINVAL when trying to create objects with
	empty names.

2009-03-13  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/mx.c: Streamline mutex support.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/task.c, src/skins/vrtx/task.c: Fix task creation.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/shadow.c, ksrc/nucleus/synch.c: Introduce debug
	option to detect mutexes held in relaxed sections.

2009-03-07  Philippe Gerum  <rpm@xenomai.org>

	* src/skins/posix/shm.c (__wrap_munmap): Forward to glibc upon EBADF.

	* ksrc/nucleus/Config.in, ksrc/arch/{powerpc, x86}/Config.in: Add
	missing SYS_STACKPOOLSZ switch.

2009-02-26  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/powerpc/patches: Update Adeos support to
	2.6.24.6-powerpc-DENX-2.4-03 (compatible with MPC5121-ADS
	hardware, see ksrc/arch/powerpc/patches/README).

2009-02-25  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/nmi.c: Rework NMI watchdog support.
	
2008-02-24  Stephane Fillod  <stephane.fillod@thomson.net>

	* src/skins/posix/shm.c (__wrap_mmap64): Make real call depend on
	configure detection (HAVE_MMAP64).
	
2009-02-24  Yi Li <liyi.dev@gmail.com>

	* configure.in: Remove -mfdpic switch. Set by default by
	bfin-linux-uclibc-gcc, and may cause error with bfin-uclinux-gcc.

2009-02-24  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/{arm, blackfin, powerpc}/hal.c (rthal_timer_request):
	Port x86 fix of race window.

2009-02-24  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/hal-common.c (rthal_timer_request): Fix race.

	* ksrc/nucleus/shadow.c (gatekeeper_thread): Do not refer to
	uninit sched slots.

2009-02-23  Paul Corner <paul_c@tuxcnc.org>

	* debian/prepare-patch.sh: Fix patch generation rules.

2009-02-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.19-x86-2.2-06,
	2.6.28.7-x86-2.2-06.

2009-02-17  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-05.

	* ksrc/arch/powerpc/patches: Upgrade to
	2.6.28.5-powerpc-DENX-2.4-04.

2009-02-13  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* src/skins/posix/shm.c (__wrap_mmap, __wrap_mmap64): Map only
	part of the underlying nucleus heap.

	* ksrc/nucleus/heap.c (xnheap_mmap): Allow partial mapping in
	user-space of a nucleus heap.

	* ksrc/nucleus/sched.c, include/nucleus/sched.h: Introduce
	sched->cpu field member.

2009-02-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pod.c (xnpod_start_thread): Run start hook for
	user-space threads as well.

2009-02-11  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/posix/shm.c (ftruncate): Remove GFP_USER hint.

2009-02-07  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/blackfin, include/asm-blackfin: Prepare for 2.6.28.

	* ksrc/arch/blackfin/patches: Upgrade to 2.6.28-1.9-00 (GIT).

	* ksrc/nucleus, include/nucleus: Introduce host tick deferral.

2009-02-02  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/syscall.c (rt_queue_write): Fix memory leak.

2009-02-01  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/heap, include/asm-generic/system.h, ksrc/skins: Fix
	auto-cleanup for shared heaps.

2009-01-31  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/drivers/testing/switchtest.c,
	  src/testsuite/switchtest/switchtest.c: Add stress mode to
	switchtest.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/ia64, include/asm-ia64: Itanium port discontinued.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-04,
	2.6.28.2-x86-2.2-05.

2009-01-27  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* include/asm-x86/bits/pod_{32, 64}.h (xnarch_leave_root): Fix FPU
	backup clobbering.
	
2009-01-27  Philippe Gerum  <rpm@eclipse.xenomai.org>

	* ksrc/nucleus, include/nucleus, ksrc/skins/posix, include/posix,
	sim, include/asm-sim: Introduce POSIX sporadic scheduling (PSS)
	class.

2009-01-21  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/arch/arm/patches: Upgrade to 2.6.28-arm-1.2-00.
	
2009-01-19  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pipe.c, ksrc/skins/native/pipe.c,
	ksrc/skins/rtai/fifo.c: Refactor callback management, fixing a
	severe latency issue, and a potential memory leak in both
	disconnect/release handlers.

	* ksrc/nucleus/pipe.c (xnpipe_disconnect, xnpipe_release):

utils/slackspot/slackspot.c
/*
 * Copyright (C) 2010 Philippe Gerum <rpm@xenomai.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * This utility parses the output of the /proc/xenomai/debug/relax
 * vfile, to get backtraces of spurious relaxes.
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <error.h>
#include <stdint.h>
#include <stdlib.h>
#include <fnmatch.h>
#include <search.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <limits.h>
#include <malloc.h>
#include <getopt.h>
#include <signal.h>
#include <cobalt/uapi/signal.h>

static const struct option base_options[] = {
	{
#define help_opt	0
		.name = "help",
		.has_arg = no_argument,
	},
#define file_opt	1
	{
		.name = "file",
		.has_arg = required_argument,
	},
#define path_opt	2
	{
		.name = "path",
		.has_arg = required_argument,
	},
#define filter_opt	3	/* Alias for filter-in */
	{
		.name = "filter",
		.has_arg = required_argument,
	},
#define filter_in_opt	4
	{
		.name = "filter-in",
		.has_arg = required_argument,
	},
#define filter_out_opt	5	/* Alias for !filter-in */
	{
		.name = "filter-out",
		.has_arg = required_argument,
	},
	{ /* Sentinel */ }
};

struct relax_spot;

struct filter {
	int (*op)(struct filter *f, struct relax_spot *p);
	char *exp;
	struct filter *next;
} *filter_list = NULL;

int filter_not = 0;

struct ldpath_dir {
	char *path;
	struct ldpath_dir *next;
} *ldpath_list = NULL;

struct location {
	unsigned long pc;
	char *function;
	char *file;
	int lineno;
	struct location *next;	/* next in mapping. */
};

struct mapping {
	char *name;
	struct location *locs;
	struct mapping *next;
} *mapping_list = NULL;

static const struct location undefined_location; /* All zero. */

struct relax_spot {
	char *exe_path;
	char *thread_name;
	char *reason;
	pid_t pid;
	int hits;
	int depth;
	struct backtrace {
		unsigned long pc;
		struct mapping *mapping;
		const struct location *where;
	} backtrace[SIGSHADOW_BACKTRACE_DEPTH];
	struct relax_spot *next;
} *spot_list = NULL;

int spot_count, filtered_count = 0;

const char *toolchain_prefix;

static int filter_thread(struct filter *f, struct relax_spot *p)
{
	return fnmatch(f->exp, p->thread_name, 0);
}

static int filter_pid(struct filter *f,  struct relax_spot *p)
{
	char pid[16];

	sprintf(pid, "%d", p->pid);

	return fnmatch(f->exp, pid, 0);
}

static int filter_exe(struct filter *f, struct relax_spot *p)
{
	return fnmatch(f->exp, p->exe_path, FNM_PATHNAME);
}

static int filter_function(struct filter *f, struct relax_spot *p)
{
	struct backtrace *b;
	int depth;

	for (depth = 0, b = p->backtrace; depth < p->depth; b++, depth++) {
		if (b->where->function &&
		    !fnmatch(f->exp, b->where->function, 0))
			return 0;
	}

	return FNM_NOMATCH;
}

static int filter_file(struct filter *f, struct relax_spot *p)
{
	struct backtrace *b;
	int depth;

	for (depth = 0, b = p->backtrace; depth < p->depth; b++, depth++) {
		if (b->where->file &&
		    !fnmatch(f->exp, b->where->file, FNM_PATHNAME))
			return 0;
	}

	return FNM_NOMATCH;
}

static int filter_map(struct filter *f, struct relax_spot *p)
{
	struct backtrace *b;
	int depth;

	for (depth = 0, b = p->backtrace; depth < p->depth; b++, depth++) {
		if (*b->mapping->name != '?' &&
		    !fnmatch(f->exp, b->mapping->name, FNM_PATHNAME))
			return 0;
	}

	return FNM_NOMATCH;
}

static int build_filter_list(const char *filters)
{
	char *filter, *name;
	struct filter *f;
	int ret;

	if (filters == NULL)
		return 0;

	filter = strdup(filters);
	while ((filter = strtok(filter, ",")) != NULL) {
		f = malloc(sizeof(*f));
		ret = sscanf(filter, "%m[a-z]=%m[^\n]", &name, &f->exp);
		if (ret != 2)
			return EINVAL;
		if (strcmp(name, "thread") == 0)
			f->op = filter_thread;
		else if (strcmp(name, "pid") == 0)
			f->op = filter_pid;
		else if (strcmp(name, "exe") == 0)
			f->op = filter_exe;
		else if (strcmp(name, "function") == 0)
			f->op = filter_function;
		else if (strcmp(name, "file") == 0)
			f->op = filter_file;
		else if (strcmp(name, "map") == 0)
			f->op = filter_map;
		else
			return EINVAL;
		f->next = filter_list;
		filter_list = f;
		filter = NULL;
	}

	return 0;
}

static inline int match_filter_list(struct relax_spot *p)
{
	struct filter *f;

	for (f = filter_list; f; f = f->next) {
		if (f->op(f, p))
			return 1 ^ filter_not;
	}

	return 0 ^ filter_not;	/* All matched. */
}

static void build_ldpath_list(const char *ldpath)
{
	char *dir, *cccmd, *search_path, *p;
	struct ldpath_dir *dpath;
	FILE *fp;
	int ret;

	if (ldpath == NULL)
		goto cross_toolchain;

	dir = strdup(ldpath);
	while ((dir = strtok(dir, ":")) != NULL) {
		dpath = malloc(sizeof(*dpath));
		dpath->path = dir;
		dpath->next = ldpath_list;
		ldpath_list = dpath;
		dir = NULL;
	}

cross_toolchain:
	if (toolchain_prefix == NULL)
		return;

	ret = asprintf(&cccmd, "%sgcc -print-search-dirs|grep '^libraries: ='",
		       toolchain_prefix);
	if (ret < 0)
		goto no_mem;

	fp = popen(cccmd, "r");
	if (fp == NULL)
		error(1, errno, "cannot run %s", cccmd);
	free(cccmd);

	ret = fscanf(fp, "libraries: =%m[^\n]\n", &search_path);
	if (ret != 1)
		goto bad_output;

	/*
	 * Feed our ldpath list with the cross-compiler's search list
	 * for libraries.
	 */
	dir = search_path;
	while ((dir = strtok(dir, ":")) != NULL) {
		p = strrchr(dir, '/');
		if (*p)
			*p = '\0';
		dpath = malloc(sizeof(*dpath));
		dpath->path = dir;
		dpath->next = ldpath_list;
		ldpath_list = dpath;
		dir = NULL;
	}

	pclose(fp);

	return;
no_mem:
	error(1, ENOMEM, "build_ldpath_list failed");

bad_output:
	error(1, 0, "garbled gcc output for -print-search-dirs");
}

static char *resolve_path(char *mapping)
{
	struct ldpath_dir *dpath;
	char *path, *basename;
	int ret;

	/*
	 * Don't use the original mapping name verbatim if
	 * CROSS_COMPILE was specified, it is unlikely that the right
	 * target file could be found at the same place on the host.
	 */
	if (*mapping == '?' ||
	    (toolchain_prefix == NULL && access(mapping, F_OK) == 0))
		return mapping;

	basename = strrchr(mapping, '/');
	if (basename++ == NULL)
		basename = mapping;

	for (dpath = ldpath_list; dpath; dpath = dpath->next) {
		ret = asprintf(&path, "%s/%s", dpath->path, basename);
		if (ret < 0)
			goto no_mem;
		/* Pick first match. */
		if (access(path, F_OK) == 0) {
			free(mapping);
			return path;
		}
		free(path);
	}

	/* No match. Leave the mapping name unchanged */
	return mapping;

no_mem:
	error(1, ENOMEM, "resolve_path failed");
	return NULL;		/* not reached. */
}

static void read_spots(FILE *fp)
{
	struct relax_spot *p;
	struct mapping *m;
	unsigned long pc;
	char *mapping, c;
	ENTRY e, *ep;
	int ret;

	ret = fscanf(fp, "%d\n", &spot_count);
	if (ret != 1) {
		if (feof(fp))
			return;
		goto bad_input;
	}

	hcreate(spot_count * SIGSHADOW_BACKTRACE_DEPTH);

	for (;;) {
		p = malloc(sizeof(*p));
		if (p == NULL)
			error(1, 0, "out of memory");

		ret = fscanf(fp, "%m[^\n]\n", &p->exe_path);
		if (ret != 1) {
			if (feof(fp))
				return;
			goto bad_input;
		}

		ret = fscanf(fp, "%d %d %m[^ ] %m[^\n]\n",
			     &p->pid, &p->hits, &p->reason, &p->thread_name);
		if (ret != 4)
			goto bad_input;

		p->depth = 0;
		for (;;) {
			if (p->depth >= SIGSHADOW_BACKTRACE_DEPTH)
				break;
			c = getc(fp);
			if (c == '.' && getc(fp) == '\n')
				break;
			ungetc(c, fp);
			ret = fscanf(fp, "%lx %m[^\n]\n", &pc, &mapping);
			if (ret != 2)
				goto bad_input;

			mapping = resolve_path(mapping);
			e.key = mapping;
			ep = hsearch(e, FIND);
			if (ep == NULL) {
				m = malloc(sizeof(*m));
				if (m == NULL)
					goto no_mem;
				m->name = mapping;
				m->locs = NULL;
				m->next = mapping_list;
				mapping_list = m;
				e.data = m;
				ep = hsearch(e, ENTER);
				if (ep == NULL)
					goto no_mem;
			} else {
				free(mapping);
				m = ep->data;
			}

			/*
			 * Move one byte backward to point to the call
			 * site, not to the next instruction. This
			 * usually works fine...
			 */
			p->backtrace[p->depth].pc = pc - 1;
			p->backtrace[p->depth].mapping = m;
			p->backtrace[p->depth].where = &undefined_location;
			p->depth++;
		}

		if (p->depth == 0)
			goto bad_input;

		p->next = spot_list;
		spot_list = p;
	}

bad_input:
	error(1, 0, "garbled trace input");
no_mem:
	error(1, ENOMEM, "read_spots failed");
}

static inline
struct location *find_location(struct location *head, unsigned long pc)
{
	struct location *l = head;

	while (l) {
		if (l->pc == pc)
			return l;
		l = l->next;
	}

	return NULL;
}

static void resolve_spots(void)
{
	char *a2l, *a2lcmd, *s, buf[BUFSIZ];
	struct relax_spot *p;
	struct backtrace *b;
	struct location *l;
	struct mapping *m;
	struct stat sbuf;
	int ret, depth;
	FILE *fp;

	/*
	 * Fill the mapping cache with one location record per
	 * distinct PC value mentioned for each mapping.  The basic
	 * idea is to exec a single addr2line instance for all PCs
	 * belonging to any given mapping, instead of one instance per
	 * call site in each and every frame. This way, we may run
	 * slackspot on low-end targets with limited CPU horsepower,
	 * without going for unreasonably long coffee breaks.
	 */
	for (p = spot_list; p; p = p->next) {
		for (depth = 0; depth < p->depth; depth++) {
			b = p->backtrace + depth;
			l = find_location(b->mapping->locs, b->pc);
			if (l) {
				/* PC found in mapping cache. */
				b->where = l;
				continue;
			}

			l = malloc(sizeof(*l));
			if (l == NULL)
				goto no_mem;

			l->pc = b->pc;
			l->function = NULL;
			l->file = NULL;
			l->lineno = 0;
			b->where = l;
			l->next = b->mapping->locs;
			b->mapping->locs = l;
		}
	}

	/*
	 * For each mapping, try resolving PC values as source
	 * locations.
	 */
	for (m = mapping_list; m; m = m->next) {
		if (*m->name == '?')
			continue;

		ret = stat(m->name, &sbuf);
		if (ret || !S_ISREG(sbuf.st_mode))
			continue;

		ret = asprintf(&a2l,
			       "%saddr2line --demangle --inlines --functions --exe=%s",
			       toolchain_prefix, m->name);
		if (ret < 0)
			goto no_mem;

		for (l = m->locs, s = a2l, a2lcmd = NULL; l; l = l->next) {
			ret = asprintf(&a2lcmd, "%s 0x%lx", s, l->pc);
			if (ret < 0)
				goto no_mem;
			free(s);
			s = a2lcmd;
		}

		fp = popen(a2lcmd, "r");
		if (fp == NULL)
			error(1, errno, "cannot run %s", a2lcmd);

		for (l = m->locs; l; l = l->next) {
			ret = fscanf(fp, "%ms\n", &l->function);
			if (ret != 1)
				goto bad_output;
			/*
			 * Don't trust fscanf range specifier, we may
			 * have colons in the pathname.
			 */
			s = fgets(buf, sizeof(buf), fp);
			if (s == NULL)
				goto bad_output;
			s = strrchr(s, ':');
			if (s == NULL)
				continue;
			*s++ = '\0';
			if (strcmp(buf, "??")) {
				l->lineno = atoi(s);
				l->file = strdup(buf);
			}
		}

		pclose(fp);
		free(a2lcmd);
	}

	return;

bad_output:
	error(1, 0, "garbled addr2line output");
no_mem:
	error(1, ENOMEM, "resolve_locations failed");
}

static inline void put_location(struct relax_spot *p, int depth)
{
	struct backtrace *b = p->backtrace + depth;
	const struct location *where = b->where;

	printf("   #%-2d 0x%.*lx ", depth, LONG_BIT / 4, where->pc);
	if (where->function)
		printf("%s() ", where->function);
	if (where->file) {
		printf("in %s", where->file);
		if (where->lineno)
			printf(":%d", where->lineno);
	} else {
		if (where->function == NULL)
			printf("??? ");
		if (*b->mapping->name != '?')
			printf("in [%s]", b->mapping->name);
	}
	putchar('\n');
}

static void display_spots(void)
{
	struct relax_spot *p;
	int depth, hits;

	for (p = spot_list, hits = 0; p; p = p->next) {
		hits += p->hits;
		if (match_filter_list(p)) {
			filtered_count++;
			continue;
		}
		printf("\nThread[%d] \"%s\" started by %s",
		       p->pid, p->thread_name, p->exe_path);
		if (p->hits > 1)
			printf(" (%d times)", p->hits);
		printf(":\n");
		printf("Caused by: %s\n", p->reason);
		for (depth = 0; depth < p->depth; depth++)
			put_location(p, depth);
	}

	if (filtered_count)
		printf("\n(%d spots filtered out)\n",
		       filtered_count);
	if (hits < spot_count)
		printf("\nWARNING: only %d/%d spots retrieved (some were lost)\n",
		       hits, spot_count);
}

static void usage(void)
{
	fprintf(stderr, "usage: slackspot [CROSS_COMPILE=<toolchain-prefix>] [options]\n");
	fprintf(stderr, "   --file <file>				use trace file\n");
	fprintf(stderr, "   --path <dir[:dir...]>			set search path for exec files\n");
	fprintf(stderr, "   --filter-in <name=exp[,name...]>		exclude non-matching spots\n");
	fprintf(stderr, "   --filter <name=exp[,name...]>		alias for --filter-in\n");
	fprintf(stderr, "   --filter-out <name=exp[,name...]>		exclude matching spots\n");
	fprintf(stderr, "   --help					print this help\n");
}

int main(int argc, char *const argv[])
{
	const char *trace_file, *filters;
	const char *ldpath;
	int c, lindex, ret;
	FILE *fp;

	trace_file = NULL;
	ldpath = NULL;
	filters = NULL;
	toolchain_prefix = getenv("CROSS_COMPILE");
	if (toolchain_prefix == NULL)
		toolchain_prefix = "";

	for (;;) {
		c = getopt_long_only(argc, argv, "", base_options, &lindex);
		if (c == EOF)
			break;
		if (c == '?') {
			usage();
			return EINVAL;
		}
		if (c > 0)
			continue;

		switch (lindex) {
		case help_opt:
			usage();
			exit(0);
		case file_opt:
			trace_file = optarg;
			break;
		case path_opt:
			ldpath = optarg;
			break;
		case filter_out_opt:
			filter_not = 1;
		case filter_in_opt:
		case filter_opt:
			filters = optarg;
			break;
		default:
			return EINVAL;
		}
	}

	fp = stdin;
	if (trace_file == NULL) {
		if (isatty(fileno(stdin))) {
			trace_file = "/proc/xenomai/debug/relax";
			goto open;
		}
	} else if (strcmp(trace_file, "-")) {
	open:
		fp = fopen(trace_file, "r");
		if (fp == NULL)
			error(1, errno, "cannot open trace file %s",
			      trace_file);
	}

	ret = build_filter_list(filters);
	if (ret)
		error(1, 0, "bad filter expression: %s", filters);

	build_ldpath_list(ldpath);
	read_spots(fp);

	if (spot_list == NULL) {
		fputs("no slacker\n", stderr);
		return 0;	/* This is not an error. */
	}

	resolve_spots();
	display_spots();

	return 0;
}
